Mathlib/GroupTheory/Coxeter/Inversion.lean

```diff
@@ -0,0 +1,443 @@
+    cs.IsLeftInversion w (s i) ↔ cs.IsLeftDescent w i := by
+  simp [IsLeftInversion, IsLeftDescent, cs.isReflection_simple i]
+
+/-- The right inversion sequence of `ω`. The right inversion sequence of a word
+$s_{i_1} \cdots s_{i_\ell}$ is the sequence
+$$s_{i_\ell}\cdots s_{i_1}\cdots s_{i_\ell}, \ldots,
+    s_{i_{\ell}}s_{i_{\ell - 1}}s_{i_{\ell - 2}}s_{i_{\ell - 1}}s_{i_\ell}, \ldots,
+    s_{i_{\ell}}s_{i_{\ell - 1}}s_{i_\ell}, s_{i_\ell}.$$
+-/
+def rightInvSeq (ω : List B) : List W :=
+  match ω with
+  | []          => []
+  | i :: ω      => (π ω)⁻¹ * (s i) * (π ω) :: rightInvSeq ω
+
+/-- The left inversion sequence of `ω`. The left inversion sequence of a word
+$s_{i_1} \cdots s_{i_\ell}$ is the sequence
+$$s_{i_1}, s_{i_1}s_{i_2}s_{i_1}, s_{i_1}s_{i_2}s_{i_3}s_{i_2}s_{i_1}, \ldots,
+    s_{i_1}\cdots s_{i_\ell}\cdots s_{i_1}.$$
+-/
+def leftInvSeq (ω : List B) : List W :=
+  match ω with
+  | []          => []
+  | i :: ω      => s i :: List.map (MulAut.conj (s i)) (leftInvSeq ω)
+
+local prefix:100 "ris" => cs.rightInvSeq
+local prefix:100 "lis" => cs.leftInvSeq
+
+@[simp] theorem rightInvSeq_nil : ris [] = [] := rfl
+
+@[simp] theorem leftInvSeq_nil : lis [] = [] := rfl
+
+@[simp] theorem rightInvSeq_singleton (i : B) : ris [i] = [s i] := by simp [rightInvSeq]
+
+@[simp] theorem leftInvSeq_singleton (i : B) : lis [i] = [s i] := rfl
+
+theorem rightInvSeq_concat (ω : List B) (i : B) :
+    ris (ω.concat i) = (List.map (MulAut.conj (s i)) (ris ω)).concat (s i) := by
+  induction' ω with j ω ih
+  · simp
+  · dsimp [rightInvSeq]
+    rw [ih]
+    simp only [concat_eq_append, wordProd_append, wordProd_cons, wordProd_nil, mul_one, mul_inv_rev,
+      inv_simple, cons_append, cons.injEq, and_true]
+    group
+
+private theorem leftInvSeq_eq_reverse_rightInvSeq_reverse (ω : List B) :
+    lis ω = (ris ω.reverse).reverse := by
+  induction' ω with i ω ih
+  · simp
+  · rw [leftInvSeq, reverse_cons, ← concat_eq_append, rightInvSeq_concat, ih]
+    simp [map_reverse]
+
+theorem leftInvSeq_concat (ω : List B) (i : B) :
+    lis (ω.concat i) = (lis ω).concat ((π ω) * (s i) * (π ω)⁻¹) := by
+  simp [leftInvSeq_eq_reverse_rightInvSeq_reverse, rightInvSeq]
+
+theorem rightInvSeq_reverse (ω : List B) :
+    ris (ω.reverse) = (lis ω).reverse := by
+  simp [leftInvSeq_eq_reverse_rightInvSeq_reverse]
+
+theorem leftInvSeq_reverse (ω : List B) :
+    lis (ω.reverse) = (ris ω).reverse := by
+  simp [leftInvSeq_eq_reverse_rightInvSeq_reverse]
+
+@[simp] theorem length_rightInvSeq (ω : List B) : (ris ω).length = ω.length := by
+  induction' ω with i ω ih
+  · simp
+  · simpa [rightInvSeq]
+
+@[simp] theorem length_leftInvSeq (ω : List B) : (lis ω).length = ω.length := by
+  simp [leftInvSeq_eq_reverse_rightInvSeq_reverse]
+
+theorem getD_rightInvSeq (ω : List B) (j : ℕ) :
+    (ris ω).getD j 1 = (π (ω.drop (j + 1)))⁻¹
+        * (Option.map (cs.simple) (ω.get? j)).getD 1
+        * π (ω.drop (j + 1)) := by
```

Comment:
I find it a little hard to scan the structure of the conclusion; maybe a line break after the `=`? Same for the next theorem